File: 2022/06/advent202206.py
file1 = open('input.txt','r')
text = file1.read()

letters = {}

for i in range(0, len(text)):
    if i >= 4:
        letters[text[i - 4]] -= 1
        if letters[text[i - 4]] == 0:
            letters.pop(text[i - 4])
    if text[i] in letters:
        letters[text[i]] += 1
    else:
        letters[text[i]] = 1
    if len(letters) == 4:
        print(i + 1)
        break

letters2 = {}
for i in range(0, len(text)):
    if i >= 14:
        letters2[text[i - 14]] -= 1
        if letters2[text[i - 14]] == 0:
            letters2.pop(text[i - 14])
    if text[i] in letters2:
        letters2[text[i]] += 1
    else:
        letters2[text[i]] = 1
    if len(letters2) == 14:
        print(i + 1)
        break

File: 2022/06/advent202206.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;

class Solver {

  private static String parseInput() {
    String fileName = "input.txt";
    String line = "";

    try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
      line = br.readLine();
    } catch (IOException e) {
      e.printStackTrace();
    }

    return line;
  }

  private static int findMarker(String signal, int distinctChars) {
    HashMap<Character, Integer> letters = new HashMap<>();
    int n = distinctChars;
    int i = 0;
    while (i < n) {
      Character c = signal.charAt(i);
      if (letters.containsKey(c)) {
        n = Math.max(n, letters.get(c) + distinctChars + 1);
      }
      letters.put(c, i);
      i++;
    }
    return i;
  }

  public static void main(String[] args) {
    String signal = parseInput();
    System.out.println(findMarker(signal, 4));
    System.out.println(findMarker(signal, 14));

  }
}
